presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainOrcPredicate.java

```diff
@@ -71,9 +91,166 @@ public boolean matches(long numberOfRows, Map<Integer, ColumnStatistics> statist
+            if (bloomfilters == null || bloomfilters.isEmpty()) {
+                // need bloom filters, treat as failure: read
+                return true;
+            }
+
+            // get domain
+            Domain domain = effectivePredicateDomains.get(columnReference.getColumn());
+            if (domain.isAll()) {
+                // all values are okay, go to next one as we don't have to check the bf
+                continue;
+            }
+
+            // extract values from domain
+            Collection<Object> predicateValues = predicateValuesFromDomain(domain);
+            if (predicateValues == null || predicateValues.isEmpty()) {
+                // no values checked, treat as failure: read
+                return true;
+            }
+
+            // run values against the bloomfilters: none can match in order to skip a read
+            for (Object o : predicateValues) {
+                for (RowGroupBloomfilter rowGroupBloomfilter : bloomfilters) {
+                    BloomFilter bloomfilter = rowGroupBloomfilter.getBloomfilter();
+
+                    TruthValue truthValue = checkInBloomFilter(bloomfilter, o, columnStatistics.getHasNull());
+                    if (truthValue == TruthValue.YES || truthValue == TruthValue.YES_NO || truthValue == TruthValue.YES_NO_NULL || truthValue == TruthValue.YES_NULL) {
+                        // bloom filter is matched here return true so we select this stripe as it likely contains data which we need to read
+                        return true;
+                    }
+                }
+            }
+        }
+
+        // none of the bloomfilters caused a "hit" meaning we should not read
+        log.debug("Not reading, didn't match any of the bloomfilters, data is not here");
+        return false;
+    }
+
+    protected Collection<Object> predicateValuesFromDomain(Domain domain)
+    {
+        Collection<Object> predicateValues = null;
+        ValueSet values = domain.getValues();
+        if (values instanceof EquatableValueSet) {
+            EquatableValueSet eqValues = (EquatableValueSet) values;
+            if (eqValues.isWhiteList()) {
+                // we can only work with values we know, not excluded blacklists because other rows might contain the data we need
+                predicateValues = values.getDiscreteValues().getValues();
+            }
+        }
+        else if (values instanceof SortedRangeSet) {
+            SortedRangeSet sortedRangeSet = (SortedRangeSet) values;
+            // sorted range set is used for integer comparison (e.g. id = 123 ) where min and max is the same value
+            if (sortedRangeSet.isSingleValue()) {
+                predicateValues = new ArrayList<>();
+                predicateValues.add(sortedRangeSet.getSingleValue());
+            }
+        }
+        return predicateValues;
+    }
+
+    private TruthValue checkInBloomFilter(BloomFilter bf, Object predObj, boolean hasNull)
```

Comment:
Why you need `TruthValue` here? Looks like you only care if there was any hit in `BloomFilter`, regardless of nullability.